shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -37,52 +37,66 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
         baseClasses.iterator.filterNot(traversed).flatMap(v =>
           ctx.tyDefs.get(v.name).fold(Set.empty[Var])(_.allBaseClasses(ctx)(traversed + v)))
     val (tparams: List[TypeName], targs: List[TypeVariable]) = tparamsargs.unzip
-    def thisTy(implicit prov: TypeProvenance): TypeRef = TypeRef(nme, targs)(prov)
+    def thisTy(prov: TypeProvenance): TypeRef = TypeRef(nme, targs)(prov)
     def wrapMethod(pt: PolymorphicType, prov: TypeProvenance): MethodType =
-      MethodType(pt.level, S((thisTy(prov), pt.body)), nme)(prov)
+      MethodType(pt.level, S((thisTy(prov), pt.body)), nme :: Nil, isInherited = false)(prov)
   }
   
-  private case class MethodDefs(
-    tn: TypeName,
-    parents: List[MethodDefs],
+  /** Represent a set of methods belonging to some owner type.
+    * This includes explicitly declared/defined as well as inherited methods. */
+  private case class MethodSet(
+    ownerType: TypeName,
+    parents: List[MethodSet],
     decls: Map[Str, MethodType],
     defns: Map[Str, MethodType],
   ) {
-    private def &(that: MethodDefs)(tn: TypeName, newdefns: Set[Str])(implicit raise: Raise): MethodDefs =
-      MethodDefs(
+    private def &(that: MethodSet)(tn: TypeName, overridden: Set[Str])(implicit raise: Raise): MethodSet =
+      MethodSet(
         tn,
         this.parents ::: that.parents,
         mergeMap(this.decls, that.decls)(_ & _),
         (this.defns.iterator ++ that.defns.iterator).toSeq.groupMap(_._1)(_._2).flatMap {
           case _ -> Nil => die
           case mn -> (defn :: Nil) => S(mn -> defn)
-          case mn -> defns if newdefns(mn) => N
+          case mn -> defns if overridden(mn) => N  // ignore a previously-defined method if it's overridden
           case mn -> defns =>
             err(msg"An overriding method definition must be given when inheriting from multiple method definitions" -> tn.toLoc
               :: msg"Definitions of method $mn inherited from:" -> N
               :: defns.iterator.map(mt => msg"• ${mt.parents.head}" -> mt.prov.loco).toList)
-            S(mn -> MethodType(defns.head.level, N, tn :: Nil)(
+            S(mn -> MethodType(defns.head.level, N, tn :: Nil, isInherited = false)(
               TypeProvenance(tn.toLoc, "inherited method declaration")))
         }
       )
-    private def addParent(prt: TypeName)(implicit ctx: Ctx): MethodDefs = {
-      val td = ctx.tyDefs(prt.name)
-      def addThis(mt: MethodType): MethodType = mt.copy(body = mt.body.map(b => b.copy(_1 = td.thisTy(mt.prov))))
-      def add(mt: MethodType): MethodType = mt.copy(parents = prt :: mt.parents)
-      copy(decls = decls.view.mapValues(addThis).mapValues(add).toMap, defns = defns.view.mapValues(addThis).toMap)
+    /** Returns a `MethodSet` of the _inherited_ methods only,
+      *   disregarding the current MethodSet's methods... */
```

Comment:
You can also call with `top = false` to get the full set of methods (i.e. where the current methods override and inherited ones). It's just never used anywhere since we need to consider them separately for subsumption checking.

Reply:
Yeah, to me this `top` parameter is an implementation detail that is only exposed for coding convenience. I'm not sure it will do the wrong thing when called with `true` though (seems it would raise wrong error messages if a method is defined several times in a class).

Also, why name this `top` without documenting what it does? Why not name it `includeCurrentClassMethods`?

Reply:
I think we can hide the `top` parameter by making this a private helper method, and only expose a method that calls this with `top = true`.

> seems it would raise wrong error messages if a method is defined several times in a class

It actually wouldn't see that a method is defined multiple times since the method declaration and definitions are stored as `Map`s. The errors are raised in `typeMethods` instead.

I named `top` `top` because something special needs to be done if this is the "top level" set of methods, and I needed a boolean parameter to represent that. I will rename it to `includeCurrentMethods` (since the current `TypeDef` can also be a trait).

Reply:
> I think we can hide the top parameter by making this a private helper method, and only expose a method that calls this with top = true.

Yeah I know but I don't think it's worth it.

Reply:
Oops I didn't see this in time 😅